runtime/drivers/file/repo.go

```diff
@@ -135,3 +137,109 @@ func (c *connection) Delete(ctx context.Context, instID, filePath string) error
 func (c *connection) Sync(ctx context.Context, instID string) error {
 	return nil
 }
+
+func (c *connection) Watch(ctx context.Context, replay bool, callback drivers.WatchCallback) error {
+	if c.Driver() != "file" {
+		return fmt.Errorf("%s repository is not supported", c.Driver())
+	}
+
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return err
+	}
+	defer watcher.Close()
+
+	fsRoot := os.DirFS(c.Root())
+
+	var dirs []string
+	var files []string
+
+	err = doublestar.GlobWalk(fsRoot, "**", func(p string, d fs.DirEntry) error {
+		if d.IsDir() {
+			dirs = append(dirs, p)
+		} else {
+			files = append(files, p)
+		}
+
+		return nil
+	})
+	if err != nil {
+		return err
+	}
+
+	if replay {
+		for _, f := range files {
+			err = callback(drivers.WatchEvent{
+				Path: filepath.Join("/", f),
+				Type: runtimev1.FileEvent_FILE_EVENT_WRITE,
+			})
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	for _, path := range dirs {
+		relativePath := filepath.Join(c.Root(), path)
+		err := watcher.Add(relativePath)
+		if err != nil {
+			return err
+		}
+	}
+
+	for {
+		select {
+		case event, ok := <-watcher.Events:
+			if !ok {
+				return nil
+			}
```

Comment:
Should this return an error?

Reply:
The channel is closed, it can be a shutdown process, should we return an error during a planned shut down?